### ISSUE TITLE
False positive `type-assertion-failure` on exhaustive match over `Mapping[str, Any] | int` (regression in 0.0.54)

### ISSUE THREAD

**Issue by martinkozle** [astral-sh/ty#3904]
### Summary

Since 0.0.54, `assert_never` in the fallthrough case of an exhaustive match started failing, but only for one ordering of the case arms.

```py
from collections.abc import Mapping
from typing import Any, assert_never


def f(a: Mapping[str, Any] | int) -> None:
    match a:
        case Mapping():
            return
        case int():
            return
        case _:
            assert_never(a)
```

[Playground](https://play.ty.dev/9759ca71-5915-4888-923f-7de7b2e1480f)

```
❮ uvx ty@0.0.53 check ty_regression.py
All checks passed!
```

```
❯ uvx ty@0.0.54 check ty_regression.py
error[type-assertion-failure]: Argument does not have asserted type `Never`
  --> ty_regression.py:12:13
   |
12 |             assert_never(a)
   |             ^^^^^^^^^^^^^-^
   |                          |
   |                          Inferred type of argument is `Mapping[str, Any] & ~Mapping[str, Any] & ~int`
   |
info: `Never` and `Mapping[str, Any] & ~Mapping[str, Any] & ~int` are not equivalent types

Found 1 diagnostic
```

```
❮ uvx ty@0.0.56 check ty_regression.py
error[type-assertion-failure]: Argument does not have asserted type `Never`
  --> ty_regression.py:12:13
   |
12 |             assert_never(a)
   |             ^^^^^^^^^^^^^-^
   |                          |
   |                          Inferred type of argument is `Mapping[str, Any] & ~Mapping[str, Any] & ~int`
   |
info: `Never` and `Mapping[str, Any] & ~Mapping[str, Any] & ~int` are not equivalent types

Found 1 diagnostic
```

Note the inferred type contains `Mapping[str, Any] & ~Mapping[str, Any]`, which looks like it should have been reduced away.

Swapping the two case arms makes the error disappear.

```py
    match a:
        case int():
            return
        case Mapping():
            return
        case _:
            assert_never(a)  # no error
```

Some things I tried while narrowing it down:

- A single-type version (`a: Mapping[str, Any]` with only `case Mapping():` before the fallthrough) does **not** reproduce it. It needs the union plus the second arm.
- `Mapping[Any, Any] | int` also reproduces it, `Sequence[Any] | int` does not.

Command was just `uvx ty@0.0.54 check ty_regression.py`, default settings, no ty config in `pyproject.toml`.

I suspect this comes from the match exhaustiveness changes in 0.0.54 (astral-sh/ruff#26283 / astral-sh/ruff#26284).

### Version

ty 0.0.54 (bf8a7dd58 2026-06-25)
